src/VCR/MatchObserver.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace VCR;
+
+class MatchObserver
+{
+    // This is an array keyed by a hash of the (first aka stored) request.
+    private $requestToMatchersMap;
+    private $requestCount0;
```

Comment:
I'm embarassed on this code. I'm sure there are better ways to organize it that are simpler and more straightforward conceptually, and also much easier to read! I will clean it up.

I agree on the should observe. The problem I had was that multiple calls to getRequest (first hasResponse uses it, then playback uses it) would record multiple mismatches. By changing the mismatch explainer to be a value / not a service, this problem will be removed: we won't need a should observe / start observing / stop observing.